tla/consensus/SIMccfraft.tla

```diff
@@ -36,7 +36,7 @@ SIMTimeout(i) ==
 \* The state constraint  StopAfter  stops TLC after the alloted
 \* time budget is up, unless TLC encounteres an error first.
 StopAfter ==
-    LET timeout == IF "SIM_TIMEOUT" \in DOMAIN IOEnv THEN atoi(IOEnv.SIM_TIMEOUT) ELSE 1200
+    LET timeout == IF ("SIM_TIMEOUT" \in DOMAIN IOEnv) /\ IOEnv.SIM_TIMEOUT # "" THEN atoi(IOEnv.SIM_TIMEOUT) ELSE 1200
```

Comment:
@lemmy there is, which is why I added the check.

```
Error evaluating expression: 'atoi("")'
tlc2.tool.EvalException: The argument of atoi should be a string, but instead it is:
```

It'd be a bit nicer if it wasn't defined altogether, but that's a faff in the ADO template, whereas `SIM_TIMEOUT: ${{parameter.timeout}}` in the `env:` is really straightforward. Alternatively we could duplicate the default in the template, but there would be two defaults.

What's be really nice is a proper programmable CLI, where a spec can declare its parameters, complete with types, range restrictions and documentation (like argparse, cli11 etc), and an ADO action that's smart enough to call --help (or a machine-readable --json-help), so that the pipeline can just say something like:

```
- script-with-help:
    path: ./tlc.sh sim.tla
    args:
      timeout: ${{parameter.timeout}}
```

with a proper optional type for parameters that would cause them not to be passed when unset. What's that beeping noise? I think it might be my alarm clock... :)

Reply:
Specs do declare their parameters (`CONSTANTS`) and one has the full expressiveness of const-level formulas to `ASSUME` constraints.   I'd argue that the problem is rather that the parsing *into* constants fails before TLC evaluates the spec's ASSUMEs.

TLC could do two passes over the ASSUMEs at startup, s.t. "CONSTANT-free" `ASSUMEs` are evaluated first.  As a workaround, we could introduce a `PRECONDITION` (analog to `POSTCONDITION`) that evaluates a const-level yet constant-free predicate.  With that and a few more helpers along the lines of `atoi`, we could assume the format of the environment variables.

Reply:
That's a piece of the puzzle, but without a --help to expose the schema of the inputs, you would only get slightly earlier/better errors. Also the auto-wrapper for ADO doesn't exist, but is probably feasible by re-using what bash_completion does. It might be a fun hackathon project.